# Preprocessing for PubMed Dataset

## 1. Load Dataset

In [21]:
# load by huggingface datasets load_dataset
from datasets import load_dataset
pubmed_copy = load_dataset('json', data_dir='../datasets/pubmed-dataset-copy')

/opt/homebrew/anaconda3/envs/summarization/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Extracting data files: 100%|██████████| 3/3 [00:00<00:00, 1081.10it/s]
Generating train split: 119924 examples [00:04, 24230.86 examples/s]
Generating validation split: 6633 examples [00:00, 25040.68 examples/s]
Generating test split: 6658 examples [00:00, 28569.08 examples/s]


In [32]:
# load by pandas
import pandas as pd
train_df = pd.read_json('../datasets/pubmed-dataset-copy/train.json', lines=True)
val_df = pd.read_json('../datasets/pubmed-dataset-copy/val.json', lines=True)
test_df = pd.read_json('../datasets/pubmed-dataset-copy/test.json', lines=True)

In [33]:
train_df

article_id                                       article_text  \
0       PMC3872579  [a recent systematic analysis showed that in 2...   
1       PMC3770628  [it occurs in more than 50% of patients and ma...   
2       PMC5330001  [tardive dystonia ( td ) , a rarer side effect...   
3       PMC4386667  [lepidoptera include agricultural pests that ,...   
4       PMC4307954  [syncope is caused by transient diffuse cerebr...   
...            ...                                                ...   
119919  PMC3502213  [eukaryotic cells depend on vesicle - mediated...   
119920  PMC3198562  [as regards the selection criteria of the post...   
119921  PMC4436536  [in most of the peer review publications in th...   
119922  PMC4251613  [the reveal registry is a longitudinal registr...   
119923  PMC3389695  [cerebral palsy is a nonprogressive central ne...   

                                            abstract_text  labels  \
0       [<S> background : the present study was carrie...     NaN   
1       [<S> backgroundanemia in patients with cancer ...     NaN   
2       [<S> tardive dystonia ( td ) is a serious side...     NaN   
3       [<S> many lepidopteran insects are agricultura...     NaN   
4       [<S> we present an unusual case of recurrent c...     NaN   
...                                                   ...     ...   
119919  [<S> long - distance trafficking of membranous...     NaN   
119920  [<S> aims and objectives : to study the stress...     NaN   
119921  [<S> abstractbackgroundthe objective of this s...     NaN   
119922  [<S> background : patients with pulmonary arte...     NaN   
119923  [<S> \n objective . to determine </S>, <S> if ...     NaN   

                                            section_names  \
0       [INTRODUCTION, MATERIALS AND METHODS, Particip...   
1       [Introduction, Patients and methods, Study des...   
2       [INTRODUCTION, CASE REPORT, DISCUSSION, Declar...   
3       [1. Introduction, 2. Insect Immunity, 3. Signa...   
4       [Introduction, Case report, Discussion, Confli...   
...                                                   ...   
119919  [Introduction, Motor-Dependent Transport of Ra...   
119920  [INTRODUCTION, MATERIALS AND METHODS, Modeling...   
119921  [Introduction, Methods, Results, Discussion, L...   
119922  [TRIAL REGISTRY:, Materials and Methods, REVEA...   
119923  [1. Introduction, 2. Method, 3. Results, 4. Di...   

                                                 sections  
0       [[a recent systematic analysis showed that in ...  
1       [[it occurs in more than 50% of patients and m...  
2       [[tardive dystonia ( td ) , a rarer side effec...  
3       [[lepidoptera include agricultural pests that ...  
4       [[syncope is caused by transient diffuse cereb...  
...                                                   ...  
119919  [[eukaryotic cells depend on vesicle - mediate...  
119920  [[fiber post systems are routinely used in res...  
119921  [[in most of the peer review publications in t...  
119922  [[], [the reveal registry is a longitudinal re...  
119923  [[cerebral palsy is a nonprogressive central n...  

[119924 rows x 6 columns]

In [34]:
val_df

article_id                                       article_text  \
0     PMC4810892  [approximately , one - third of patients with ...   
1     PMC4852153  [there is an epidemic of stroke in low and mid...   
2     PMC5307612  [\n cardiovascular diseases account for the hi...   
3     PMC3309663  [results of a liquid culturing system ( bd bac...   
4     PMC4413883  [the need for magnetic resonance imaging ( mri...   
...          ...                                                ...   
6628  PMC3934811  [a 48-year - old woman with a positive brca1 g...   
6629  PMC4503916  [with an increasing number of laparoscopic cho...   
6630  PMC4754593  [upton and mccomas first reported double crush...   
6631  PMC5317293  [acute respiratory distress syndrome ( ards ) ...   
6632  PMC4672074  [cardiac resynchronization therapy ( crt ) is ...   

                                          abstract_text  labels  \
0     [<S> background and aim : there is lack of sub...     NaN   
1     [<S> backgroundthe questionnaire for verifying...     NaN   
2     [<S> \n background : timely access to cardiova...     NaN   
3     [<S> to determine differences in the ability o...     NaN   
4     [<S> aimsour aim was to evaluate the potential...     NaN   
...                                                 ...     ...   
6628  [<S> a 48-year - old woman with a positive brc...     NaN   
6629  [<S> we present a case on abdominal wall absce...     NaN   
6630  [<S> double compression of the ulnar nerve , i...     NaN   
6631  [<S> backgroundno definitive conclusions have ...     NaN   
6632  [<S> backgroundvarious difficulties can occur ...     NaN   

                                          section_names  \
0     [Introduction, Subjects and Methods, Results, ...   
1     [1. Introduction, 2. Methods, 2.1. Study sites...   
2     [Introduction, Methods, Results, Discussion, C...   
3      [The Study, Conclusions, Supplementary Material]   
4     [Introduction, Methods, Patient selection, Mag...   
...                                                 ...   
6628              [Case Report, Discussion, Conclusion]   
6629  [INTRODUCTION, CASE REPORT, DISCUSSION, CONFLI...   
6630            [INTRODUCTION, CASE REPORT, DISCUSSION]   
6631  [Background, Material and Methods, Patient sel...   
6632  [Introduction, Material and methods, Study pat...   

                                               sections  
0     [[venous thromboembolism ( vte ) comprising of...  
1     [[there is an epidemic of stroke in low and mi...  
2     [[\n cardiovascular diseases account for the h...  
3     [[results of a liquid culturing system ( bd ba...  
4     [[the need for magnetic resonance imaging ( mr...  
...                                                 ...  
6628  [[a 48-year - old woman with a positive brca1 ...  
6629  [[with an increasing number of laparoscopic ch...  
6630  [[upton and mccomas first reported double crus...  
6631  [[acute respiratory distress syndrome ( ards )...  
6632  [[cardiac resynchronization therapy ( crt ) is...  

[6633 rows x 6 columns]

In [35]:
test_df

article_id                                       article_text  \
0     PMC5075302  [anxiety affects quality of life in those livi...   
1     PMC3309138  [small non - coding rnas are transcribed into ...   
2     PMC4086000  [ohss is a serious complication of ovulation i...   
3     PMC3603086  [congenital adrenal hyperplasia ( cah ) refers...   
4     PMC4414990  [type 1 diabetes ( t1d ) results from the dest...   
...          ...                                                ...   
6653  PMC2817809  [neurocysticercosis is the most common parasit...   
6654  PMC5421177  [most chd prevalence data are based on populat...   
6655  PMC3336056  [renal transplantation rates are low among pat...   
6656  PMC4461384  [acute kidney injury ( aki ) , defined as an a...   
6657  PMC4637496  [human brain plasticity or neuroplasticity ref...   

                                          abstract_text  labels  \
0     [<S> research on the implications of anxiety i...     NaN   
1     [<S> small non - coding rnas include sirna , m...     NaN   
2     [<S> objective : to evaluate the efficacy and ...     NaN   
3     [<S> congenital adrenal hyperplasia is a group...     NaN   
4     [<S> objective(s):pentoxifylline is an immunom...     NaN   
...                                                 ...     ...   
6653  [<S> cysticercosis is an infection with larval...     NaN   
6654  [<S> background : congenital heart disease ( c...     NaN   
6655  [<S> we describe a successful living donor liv...     NaN   
6656  [<S> objective : the aim was to investigate th...     NaN   
6657  [<S> accumulating evidence has indicated that ...     NaN   

                                          section_names  \
0     [1. Introduction, 2. Methods, 3. Results, 4. D...   
1     [Introduction, Aberrant Expression of miRNA in...   
2     [Introduction, Materials and Methods, Results,...   
3                                          [I, M, R, D]   
4     [Introduction, Materials and Methods, Drug and...   
...                                                 ...   
6653  [1. Introduction, 2. Materials and Methods, 3....   
6654  [I, M, Subjects, Ethical approval, Protocol, D...   
6655            [Introduction, Case report, Discussion]   
6656  [INTRODUCTION, METHODS, Search Strategy, Study...   
6657  [1. Introduction, 2. Materials and Methods, 3....   

                                               sections  
0     [[anxiety affects quality of life in those liv...  
1     [[small non - coding rnas are transcribed into...  
2     [[ohss is a serious complication of ovulation ...  
3     [[congenital adrenal hyperplasia ( cah ) refer...  
4     [[type 1 diabetes ( t1d ) results from the des...  
...                                                 ...  
6653  [[neurocysticercosis is the most common parasi...  
6654  [[in developing countries , congenital heart d...  
6655  [[renal transplantation rates are low among pa...  
6656  [[acute kidney injury ( aki ) , defined as an ...  
6657  [[human brain plasticity or neuroplasticity re...  

[6658 rows x 6 columns]

In [36]:
test_df['section_names']

0       [1. Introduction, 2. Methods, 3. Results, 4. D...
1       [Introduction, Aberrant Expression of miRNA in...
2       [Introduction, Materials and Methods, Results,...
3                                            [I, M, R, D]
4       [Introduction, Materials and Methods, Drug and...
                              ...                        
6653    [1. Introduction, 2. Materials and Methods, 3....
6654    [I, M, Subjects, Ethical approval, Protocol, D...
6655              [Introduction, Case report, Discussion]
6656    [INTRODUCTION, METHODS, Search Strategy, Study...
6657    [1. Introduction, 2. Materials and Methods, 3....
Name: section_names, Length: 6658, dtype: object

* Pubmed statistic:
  | Split | samples | 
  | --- | --- |
  | Train | 119924 | 
  | Val | 6633 |
  | Test | 6658 |

* a sample of vanilla dataset
  * labels seems to used for classification task, the labels of vanilla datasets are all None, i didn't do any processing.
  
  | article_id | article_text | abstract_text | labels | section_names | sections |
  | --- | --- | --- | --- | --- | --- |
  | str | List[str] | List[str] | None | List[str] | List[List[str]] |
  | 'PMC3872579' | split by sentence<br>(似乎按照句子进行分割了, 如果要使用需要进行join) | ['<BOS> background : the present study was carried out to assess the effects of community nutrition intervention based on advocacy approach on malnutrition status among school - aged children in shiraz , iran.materials and methods : this case - control nutritional intervention has been done between 2008 and 2009 on 2897 primary and secondary school boys and girls ( 7 - 13 years old ) based on advocacy approach in shiraz , iran . </EOS>', <br> '<BOS> the project provided nutritious snacks in public schools over a 2-year period along with advocacy oriented actions in order to implement and promote nutritional intervention . for evaluation of effectiveness of the intervention growth monitoring indices of pre- and post - intervention were statistically compared.results:the frequency of subjects with body mass index lower than 5% decreased significantly after intervention among girls ( p = 0.02 ) . </EOS>', <br> '<BOS> however , there were no significant changes among boys or total population . </EOS>', <br> '<BOS> the mean of all anthropometric indices changed significantly after intervention both among girls and boys as well as in total population . </EOS>', <br> "<BOS> the pre- and post - test education assessment in both groups showed that the student 's average knowledge score has been significantly increased from 12.5  3.2 to 16.8  4.3 ( p < 0.0001).conclusion : this study demonstrates the potential success and scalability of school feeding programs in iran . </EOS>", <br> '<BOS> community nutrition intervention based on the advocacy process model is effective on reducing the prevalence of underweight specifically among female school aged children . </EOS>'] <br> 分割方式不明, 需要进行join并重新分割 | None | ['INTRODUCTION', <br>'MATERIALS AND METHODS', <br>'Participants'Instruments', <br>'Procedure', <br>'First step', <br>'Second step', <br>'Third step', <br>'Forth step', <br>'Interventions', <br>'Fifth step (assessment)', <br>'Data analysis', <br>'RESULTS', <br>'DISCUSSION', <br>'CONCLUSION'] <br><br>请注意METHODS包含了从Participants'Instruments到Data analysis的部分 | [<br>section[seq1, seq2, ...], <br>section[seq1, seq2, ...], <br>...] |
  | reformat之后的格式 | --- | --- | --- | --- | --- |
  | | | | | | List[str]|


## 2. Dataset Preprocessing

### 2.1 Keyword Table
The summary(abstract) of pubmed dataset have not been split into multiple pieces, but luckily, it has section strutures so we could mannually process it by keyword matching.
> reference: [GovReport](https://arxiv.org/pdf/2104.02112v2.pdf)
> see page 18

In [456]:
segmentation_keyword_table = {
    # Introduction and Literature
    'part_1': ['introduction', 'case', 'objectives', 'purposes', 
               'objective', 'purpose', 'background', 'literature',],
    
    # Methods
    'part_2': ['material and methods',
               'materials and methods', 'methods', 'techniques', 'methodology',
               'materials', 'research design', 'study design'],
    
    # Results
    'part_3': ['result', 'results', 'experiments', 'observations'],
    
    # Discussion and Conlusion
    'part_4': ['discussion', 'limitation', 'conclusions', 
               'conclusion', 'concluding', 'comment', 'comments', 
               'summary', 'concluding remarks'],
    
}

## 2.2 Sections

### 2.2.1 Concat sentences into one setence <br>(每一行是一个完整的sections, 由多个section组成, 其中每个section有多个句子, 需要进行合并)

In [14]:
train_df['sections'][0]

[['a recent systematic analysis showed that in 2011 , 314 ( 296 - 331 ) million children younger than 5 years were mildly , moderately or severely stunted and 258 ( 240 - 274 ) million were mildly , moderately or severely underweight in the developing countries .',
  'in iran a study among 752 high school girls in sistan and baluchestan showed prevalence of 16.2% , 8.6% and 1.5% , for underweight , overweight and obesity , respectively .',
  'the prevalence of malnutrition among elementary school aged children in tehran varied from 6% to 16% .',
  'anthropometric study of elementary school students in shiraz revealed that 16% of them suffer from malnutrition and low body weight .',
  'snack should have 300 - 400 kcal energy and could provide 5 - 10 g of protein / day . nowadays , school nutrition programs are running as the national programs , world - wide . national school lunch program in the united states',
  'there are also some reports regarding school feeding programs in developi

In [39]:
def sec_join(df):
    # 每一行
    sample_list = []
    for sample in df['sections']:
        # 每一个sample的每一个section
        sections = []
        for section in sample: 
            temp = " ".join(section)
            sections.append(temp)
        
        sample_list.append(sections)
    
    return sample_list

In [40]:
train_df['sections'] = sec_join(train_df)
test_df['sections'] = sec_join(test_df)
val_df['sections'] = sec_join(val_df)

In [42]:
pubmed_copy['train']['sections'][2]

[['tardive dystonia ( td ) , a rarer side effect after longer exposure to antipsychotics , is characterized by local or general , sustained , involuntary contraction of a muscle or muscle group , with twisting movements , generally slow , which may affect the limbs , trunk , neck , or face .',
  'td has been shown to develop in about 3% of patients who have had long - term exposure to antipsychotics .',
  '. the low risk of td for atypical antipsychotics is thought to result from their weak affinity for dopamine receptors . compared with typical ,',
  'atypical antipsychotic agents have a greater affinity for serotonin 5-ht2a than dopamine d2 receptors , with a low propensity to induce td . among this olanzapine',
  'is thought to have preferential action at mesolimbic over nigrostriatal dopaminergic pathways and is , therefore , associated with a very low incidence of extrapyramidal symptom ( eps ) .',
  'furthermore , a retrospective analysis of controlled multicentric trials suggest

In [43]:
train_df['sections'][2]

['tardive dystonia ( td ) , a rarer side effect after longer exposure to antipsychotics , is characterized by local or general , sustained , involuntary contraction of a muscle or muscle group , with twisting movements , generally slow , which may affect the limbs , trunk , neck , or face . td has been shown to develop in about 3% of patients who have had long - term exposure to antipsychotics . . the low risk of td for atypical antipsychotics is thought to result from their weak affinity for dopamine receptors . compared with typical , atypical antipsychotic agents have a greater affinity for serotonin 5-ht2a than dopamine d2 receptors , with a low propensity to induce td . among this olanzapine is thought to have preferential action at mesolimbic over nigrostriatal dopaminergic pathways and is , therefore , associated with a very low incidence of extrapyramidal symptom ( eps ) . furthermore , a retrospective analysis of controlled multicentric trials suggested that olanzapine also im

In [45]:
train_df['sections'][0]

["a recent systematic analysis showed that in 2011 , 314 ( 296 - 331 ) million children younger than 5 years were mildly , moderately or severely stunted and 258 ( 240 - 274 ) million were mildly , moderately or severely underweight in the developing countries . in iran a study among 752 high school girls in sistan and baluchestan showed prevalence of 16.2% , 8.6% and 1.5% , for underweight , overweight and obesity , respectively . the prevalence of malnutrition among elementary school aged children in tehran varied from 6% to 16% . anthropometric study of elementary school students in shiraz revealed that 16% of them suffer from malnutrition and low body weight . snack should have 300 - 400 kcal energy and could provide 5 - 10 g of protein / day . nowadays , school nutrition programs are running as the national programs , world - wide . national school lunch program in the united states there are also some reports regarding school feeding programs in developing countries . in vietnam 

### 2.2.2 删除空行（包括None）

* 删除labels列, 原数据集里就是全空值

In [46]:
# 首先需要删除labels列, 原数据集里面的labels列是空的
train_df.drop(columns=['labels'], inplace=True)
val_df.drop(columns=['labels'], inplace=True)
test_df.drop(columns=['labels'], inplace=True)

* 再删除有空值的行,确保没有问题

In [49]:
# 删除有空值的行
train_df.dropna(inplace=True)
val_df.dropna(inplace=True)
test_df.dropna(inplace=True)

* 过滤掉文章本身为空的行

In [64]:
train_df = train_df[train_df['article_text'].apply(lambda x: x != [""])]

val_df = val_df[val_df['article_text'].apply(lambda x: x != [""])]
test_df = test_df[test_df['article_text'].apply(lambda x: x != [""])]

In [80]:
# 似乎只有train数据集里有空的article_text
train_df

article_id                                       article_text  \
0       PMC3872579  [a recent systematic analysis showed that in 2...   
1       PMC3770628  [it occurs in more than 50% of patients and ma...   
2       PMC5330001  [tardive dystonia ( td ) , a rarer side effect...   
3       PMC4386667  [lepidoptera include agricultural pests that ,...   
4       PMC4307954  [syncope is caused by transient diffuse cerebr...   
...            ...                                                ...   
119919  PMC3502213  [eukaryotic cells depend on vesicle - mediated...   
119920  PMC3198562  [as regards the selection criteria of the post...   
119921  PMC4436536  [in most of the peer review publications in th...   
119922  PMC4251613  [the reveal registry is a longitudinal registr...   
119923  PMC3389695  [cerebral palsy is a nonprogressive central ne...   

                                            abstract_text  \
0       [<S> background : the present study was carrie...   
1       [<S> backgroundanemia in patients with cancer ...   
2       [<S> tardive dystonia ( td ) is a serious side...   
3       [<S> many lepidopteran insects are agricultura...   
4       [<S> we present an unusual case of recurrent c...   
...                                                   ...   
119919  [<S> long - distance trafficking of membranous...   
119920  [<S> aims and objectives : to study the stress...   
119921  [<S> abstractbackgroundthe objective of this s...   
119922  [<S> background : patients with pulmonary arte...   
119923  [<S> \n objective . to determine </S>, <S> if ...   

                                            section_names  \
0       [INTRODUCTION, MATERIALS AND METHODS, Particip...   
1       [Introduction, Patients and methods, Study des...   
2       [INTRODUCTION, CASE REPORT, DISCUSSION, Declar...   
3       [1. Introduction, 2. Insect Immunity, 3. Signa...   
4       [Introduction, Case report, Discussion, Confli...   
...                                                   ...   
119919  [Introduction, Motor-Dependent Transport of Ra...   
119920  [INTRODUCTION, MATERIALS AND METHODS, Modeling...   
119921  [Introduction, Methods, Results, Discussion, L...   
119922  [TRIAL REGISTRY:, Materials and Methods, REVEA...   
119923  [1. Introduction, 2. Method, 3. Results, 4. Di...   

                                                 sections  
0       [a recent systematic analysis showed that in 2...  
1       [it occurs in more than 50% of patients and ma...  
2       [tardive dystonia ( td ) , a rarer side effect...  
3       [lepidoptera include agricultural pests that ,...  
4       [syncope is caused by transient diffuse cerebr...  
...                                                   ...  
119919  [eukaryotic cells depend on vesicle - mediated...  
119920  [fiber post systems are routinely used in rest...  
119921  [in most of the peer review publications in th...  
119922  [, the reveal registry is a longitudinal regis...  
119923  [cerebral palsy is a nonprogressive central ne...  

[117112 rows x 5 columns]

### 2.2.3 根据关键字列表合并sections

* 首先需要取section_names的每一个句子的小写
* 获取到section_names的中和keyword tabel相匹配的每一个section

In [89]:
train_df['sections'][0]

["a recent systematic analysis showed that in 2011 , 314 ( 296 - 331 ) million children younger than 5 years were mildly , moderately or severely stunted and 258 ( 240 - 274 ) million were mildly , moderately or severely underweight in the developing countries . in iran a study among 752 high school girls in sistan and baluchestan showed prevalence of 16.2% , 8.6% and 1.5% , for underweight , overweight and obesity , respectively . the prevalence of malnutrition among elementary school aged children in tehran varied from 6% to 16% . anthropometric study of elementary school students in shiraz revealed that 16% of them suffer from malnutrition and low body weight . snack should have 300 - 400 kcal energy and could provide 5 - 10 g of protein / day . nowadays , school nutrition programs are running as the national programs , world - wide . national school lunch program in the united states there are also some reports regarding school feeding programs in developing countries . in vietnam 

In [90]:
train_df['abstract_text'][0]

['<S> background : the present study was carried out to assess the effects of community nutrition intervention based on advocacy approach on malnutrition status among school - aged children in shiraz , iran.materials and methods : this case - control nutritional intervention has been done between 2008 and 2009 on 2897 primary and secondary school boys and girls ( 7 - 13 years old ) based on advocacy approach in shiraz , iran . </S>',
 '<S> the project provided nutritious snacks in public schools over a 2-year period along with advocacy oriented actions in order to implement and promote nutritional intervention . for evaluation of effectiveness of the intervention growth monitoring indices of pre- and post - intervention were statistically compared.results:the frequency of subjects with body mass index lower than 5% decreased significantly after intervention among girls ( p = 0.02 ) . </S>',
 '<S> however , there were no significant changes among boys or total population . </S>',
 '<S> 

In [457]:
import re

def keyword_matching_and_re_section(dataset):
    for idx, sample in dataset.iterrows():
        # reset for each sample
        sections = [
            [], # part_1
            [], # part_2
            [], # part_3
            [], # part_4
        ]
        section_names = [
            [], # part_1
            [], # part_2
            [], # part_3
            [], # part_4        
        ]
        # print(f"{sample['section_names']=}")
        for sec_name, sec in zip(sample['section_names'], sample['sections']):
            # print(f'{sec_name.lower()=}')
            for id, values in enumerate(segmentation_keyword_table.values()):
                # print(f'{sec_name.lower()=}, {values=}')
                if any(keyword.lower() in sec_name.lower() for keyword in values):
                    sections[id].append(sec)
                    # print(f'{sections[id]=}')
        for id, value in enumerate(sections):
            sections[id] = " ".join(sections[id])

        dataset['sections'].loc[idx] = sections
        # print(f"{dataset['sections'].iloc[idx]=}")    
        # print('------------------')

In [458]:
train_df_copy = train_df.copy()
keyword_matching_and_re_section(train_df_copy)

val_df_copy = val_df.copy()
keyword_matching_and_re_section(val_df_copy)

test_df_copy = test_df.copy()
keyword_matching_and_re_section(test_df_copy)

* 如果处理完后的sections仍然不包含完整的四段式结构, 就删除那一行

In [481]:
train_df_copy_copy = train_df_copy.copy()
train_df_copy_copy = train_df_copy_copy[train_df_copy_copy['sections'].apply(lambda x: all(s != '' for s in x))]

val_df_copy_copy = val_df_copy.copy()
val_df_copy_copy = val_df_copy_copy[val_df_copy_copy['sections'].apply(lambda x: all(s != '' for s in x))]

test_df_copy_copy = test_df_copy.copy()
test_df_copy_copy = test_df_copy_copy[test_df_copy_copy['sections'].apply(lambda x: all(s != '' for s in x))]

* 处理之后分别剩下了51489/2858/2982个样本

* 暂时进行一次保存

In [498]:
train_df_copy_copy.to_json(f'../datasets/pubmed-dataset-processed/train.json', orient='records', lines=True) # orient
val_df_copy_copy.to_json(f'../datasets/pubmed-dataset-processed/val.json', orient='records', lines=True)
test_df_copy_copy.to_json(f'../datasets/pubmed-dataset-processed/test.json', orient='records', lines=True)

In [503]:
index = 189

In [504]:
train_df_copy.iloc[index]['section_names']

['1. Introduction',
 '2. Materials and Methods',
 '3. Results',
 '4. Discussion',
 '5. Clinical Significance of BG Morphology',
 '6. Conclusions']

In [505]:
train_df_copy.iloc[index]['sections']

['bicipital groove ( bg ) is an indentation on the anterior aspect of proximal part of humerus . this groove allows tendon of long head of biceps brachi muscle enveloped in synovial sheath and ascending branch of anterior circumflex humeral artery to pass through it . it is bounded medially by lesser tubercle , laterally by greater tubercle , and superiorly by bridging of transverse humeral ligament /muscle fibers of subscapularis , supraspinatus and pectoralis major muscles [ 2 , 3 ] . this groove with transverse humeral ligament / muscle fibers bridging it provides stability and smooth functioning of tendon of long head of biceps brachi muscle and prevents its subluxation during multidirectional biomechanical movements of arms . apart from this , the greater function of biceps brachi muscle whose tendon is enshrined in bicipital groove is suppination , flexion , and screwing biomechanical movements . on motion of humerus , the proximal humerus moves in relation to fixed biceps tendon

In [506]:
train_df.iloc[index]['sections']

['bicipital groove ( bg ) is an indentation on the anterior aspect of proximal part of humerus . this groove allows tendon of long head of biceps brachi muscle enveloped in synovial sheath and ascending branch of anterior circumflex humeral artery to pass through it . it is bounded medially by lesser tubercle , laterally by greater tubercle , and superiorly by bridging of transverse humeral ligament /muscle fibers of subscapularis , supraspinatus and pectoralis major muscles [ 2 , 3 ] . this groove with transverse humeral ligament / muscle fibers bridging it provides stability and smooth functioning of tendon of long head of biceps brachi muscle and prevents its subluxation during multidirectional biomechanical movements of arms . apart from this , the greater function of biceps brachi muscle whose tendon is enshrined in bicipital groove is suppination , flexion , and screwing biomechanical movements . on motion of humerus , the proximal humerus moves in relation to fixed biceps tendon

In [507]:
train_df.iloc[index]['abstract_text']

['<S> the variant morphometry of bicipital groove is reported to be associated with pathologies of biceps tendon and is useful in surgical procedures in this region . </S>',
 '<S> the pathologies of biceps tendon are frequent causes of shoulder pain . therefore , under the condition of paucity of data pertaining to north indians , not only morphometric analysis of bicipital groove and a new definition of narrow / shallow groove to provide logical explanation for dependence of pathologies of biceps tendon on groove morphology is done but also a review of the literature has been carried out . </S>',
 '<S> various dimensions such as lengths of medial and lateral walls , width , depth , medial wall , and opening angles including incidence of supratubercular ridge of bicipital groove from 101 humerii are 23  5 , 32  5 , 8  2 , 6  1 , 48.91  10.31 , 82.20  22.62 , and 37% , respectively . </S>',
 '<S> the average height along with average width of biceps tendon and average width along with a

### 2.2.4 对abstract进行分割
* 特别注意：abstract有一部分文本需要再进行分词操作！！！！！！！

* 加载数据集

In [508]:
train_df = pd.read_json('../datasets/pubmed-dataset-processed/train.json', lines=True)
val_df = pd.read_json('../datasets/pubmed-dataset-processed/val.json', lines=True)
test_df = pd.read_json('../datasets/pubmed-dataset-processed/test.json', lines=True)

In [511]:
test_df

article_id                                       article_text  \
0     PMC5075302  [anxiety affects quality of life in those livi...   
1     PMC4086000  [ohss is a serious complication of ovulation i...   
2     PMC4414990  [type 1 diabetes ( t1d ) results from the dest...   
3     PMC5094872  [determinar a presena de anticorpos ige especf...   
4     PMC4262794  [development of human societies and industrial...   
...          ...                                                ...   
2977  PMC4325121  [if left untreated , ureteropelvic junction ob...   
2978  PMC3642939  [the hydatid cyst is a zoonosis caused by adul...   
2979  PMC2817809  [neurocysticercosis is the most common parasit...   
2980  PMC4461384  [acute kidney injury ( aki ) , defined as an a...   
2981  PMC4637496  [human brain plasticity or neuroplasticity ref...   

                                          abstract_text  \
0     [<S> research on the implications of anxiety i...   
1     [<S> objective : to evaluate the efficacy and ...   
2     [<S> objective(s):pentoxifylline is an immunom...   
3     [<S> abstractobjective : to determine the pres...   
4     [<S> background and objective :   anxiety and ...   
...                                                 ...   
2977  [<S> purposeto evaluate changes in differentia...   
2978  [<S> hydatid disease is caused by echinococcus...   
2979  [<S> cysticercosis is an infection with larval...   
2980  [<S> objective : the aim was to investigate th...   
2981  [<S> accumulating evidence has indicated that ...   

                                          section_names  \
0     [1. Introduction, 2. Methods, 3. Results, 4. D...   
1     [Introduction, Materials and Methods, Results,...   
2     [Introduction, Materials and Methods, Drug and...   
3     [Objetivo:, Mtodos:, Resultados:, Concluses:, ...   
4     [INTRODUCTION , METHODS , ETHICAL CONSIDERATIO...   
...                                                 ...   
2977  [INTRODUCTION, MATERIALS AND METHODS, RESULTS,...   
2978  [Introduction, Methods, Results, Discussion, C...   
2979  [1. Introduction, 2. Materials and Methods, 3....   
2980  [INTRODUCTION, METHODS, Search Strategy, Study...   
2981  [1. Introduction, 2. Materials and Methods, 3....   

                                               sections  
0     [anxiety affects quality of life in those livi...  
1     [ohss is a serious complication of ovulation i...  
2     [type 1 diabetes ( t1d ) results from the dest...  
3     [\n staphylococcus aureus is a gram - positive...  
4     [development of human societies and industrial...  
...                                                 ...  
2977  [if left untreated , ureteropelvic junction ob...  
2978  [the hydatid cyst is a zoonosis caused by adul...  
2979  [neurocysticercosis is the most common parasit...  
2980  [acute kidney injury ( aki ) , defined as an a...  
2981  [human brain plasticity or neuroplasticity ref...  

[2982 rows x 5 columns]

# 3. 保存数据集

In [ ]:
train_df.to_json('train.json', orient='records', lines=True)
test_df.to_json('test.json', orient='records', lines=True)
val_df.to_json('val.json', orient='records', lines=True)